### DISCUSSION TITLE
The extracted color images turn black, which happens only in a few literature journals.

### DISCUSSION THREAD

**Discussion by ZJZVZS** [pymupdf/PyMuPDF#3709] [Looking for help]
### Description of the bug

When extracting images from the literature, some journals may convert the color images into formats like 1.pdf, 2.pdf, 3.pdf. How can this be resolved?

### How to reproduce the bug

[example.pdf](https://github.com/user-attachments/files/16323369/example.pdf)
![Uploading 2.jpeg…]()
![Uploa
![Uploading 3.jpeg…]()
ding 1.jpeg…]()
[Uploading code.py…]()

### PyMuPDF version

1.24.7

### Operating system

Windows

### Python version

3.10

**Comment by JorjMcKie** [comment]
There is no evidence for what the bug is.
Apart from the the PDF itself the other links do not point to anywhere meaningful.
It overall looks like a Discussions post. Therefore transferring ...

**Comment by JorjMcKie** [comment]
Please describe again precisely what you are trying to do and what the problem is.
This includes the script's source and one example of a failing image extraction.

**Reply by JorjMcKie** [reply]
Even better: look in this repo to find examples that successfully extract images https://github.com/pymupdf/PyMuPDF-Utilities/tree/master/examples/extract-images

**Reply by ZJZVZS** [reply]
Thank you for your reply. Let me describe my problem in detail: I want to extract images from a PDF file, but for some images, such as the file (bad.jpeg), the extracted result has a black background, which is not favorable for me. I am not clear about the reasons for this issue, especially compared to successful cases like success.png. I have also attached the PDF that can be extracted successfully (success.pdf) and the PDF that failed to extract (bad.pdf), as well as my code (code.py).
![bad](https://github.com/user-attachments/assets/c7840b9a-7ed0-4d70-b4f6-a4460c066f8f)
[bad.pdf](https://github.com/user-attachments/files/16324570/bad.pdf)
[success.pdf](https://github.com/user-attachments/files/16324571/success.pdf)
![success](https://github.com/user-attachments/assets/e7d88e4b-f4a9-4a0a-b060-99a3a83962c5)
this is my code
import fitz  # PyMuPDF
import os
import re
from PIL import Image
import io

def extract_images_from_pdf(pdf_path, keyword):
    pdf_document = fitz.open(pdf_path)
    pdf_name = os.path.basename(pdf_path).replace('.pdf', '')
    output_folder = f"./{pdf_name}"
    os.makedirs(output_folder, exist_ok=True)

    print(f"Processing '{pdf_path}'...")  

    for page_number in range(len(pdf_document)):
        page = pdf_document[page_number]
        text = page.get_text("text")
        images = page.get_images(full=True)

        print(f"  Processing page {page_number + 1} of {len(pdf_document)}...")

        if re.search(keyword, text, re.IGNORECASE):
            for img_index, img in enumerate(images):
                xref = img[0]
                base_image = pdf_document.extract_image(xref)
                image_bytes = base_image["image"]
                image_extension = base_image["ext"]

                
                image = Image.open(io.BytesIO(image_bytes))

                
                if image.mode == 'RGBA':
                    
                    background = Image.new("RGB", image.size, (255, 255, 255))
                    
                    background.paste(image, (0, 0), image)
                    image = background
                elif image.mode != 'RGB':
                    
                    image = image.convert('RGB')

                
                image_filename = os.path.join(output_folder, f"image_{page_number + 1}_{img_index + 1}.{image_extension}")
                image.save(image_filename)

    pdf_document.close()
    print(f"Completed processing '{pdf_path}'.\n")  


pdf_directory = "/home/zjz/extract-image/6-test"  
keyword = "FE"  

pdf_files = [os.path.join(pdf_directory, f) for f in os.listdir(pdf_directory) if f.endswith('.pdf')]

for pdf_file in pdf_files:
    extract_images_from_pdf(pdf_file, keyword)

**Comment by JorjMcKie** [comment]
Did you look at the script I pointed to at all?
It extracts 8 flawless images from `bad.pdf`.
```python
py extract-from-pages.py bad.pdf
None
8 images in total
8 images extracted
total time 1.50535 sec
```

**Reply by JorjMcKie** [reply]
e.g. this one 
![image](https://github.com/user-attachments/assets/3f991e77-ce7a-4c7d-9499-38f1bfa6b4be)

**Reply by ZJZVZS** [reply]
Thank you very much for your response！Just now, I tried the script you provided, and it successfully extracted images from the PDF. However, I still encountered a new issue: for the PDF mentioned below, the extracted images were surprisingly split into very small parts. From my observation, it seems that the images in the PDF are also stitched together. Is there an effective way to solve this problem?
![image](https://github.com/user-attachments/assets/c8b10ee1-47d3-43a0-a6cb-8833ddb1e381)
[example.pdf](https://github.com/user-attachments/files/16324708/example.pdf)
I believe this is due to the PDF itself, but I still want to know if there is a way to resolve this issue.

**Reply by JorjMcKie** [reply]
No, there is not. It indeed is done so in the PDF. The images are **_not "stitched"_** together - just located in each other's vicinity.
You **_could_** extract the areas (boundary boxes) of the images and join rectangles into one area if they overlap.
Then do not extract images, but make a pixmap from that area and save it as an image: `pix = page.get_pixmap(clip=area)` and `pix.save("some.png")`.
You obviously need a different script for doing this.
